### PR TITLE
Fix "make install-tests" when srcdir != builddir, fix valgrind-tests

### PR DESCRIPTION
**Pull request by mbiebl**
Follow-up for 4f8425b8d5a3fb2e5ec24b77f1a7a95db55f532c and PR #5339

@martinpitt please have a look

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -6702,13 +6702,13 @@ tests += \
 install-tests: $(tests) $(TEST_DATA_FILES)
 	for f in $(tests); do \
 	    if [ -x .libs/$$f ]; then \
-		install -D -m 755 .libs/$$f $(DESTDIR)/$(testsdir)/$$f; \
+		install -D -m 755 $(top_builddir)/.libs/$$f $(DESTDIR)/$(testsdir)/$$f; \
 	    else \
-		install -D -m 755 $$f $(DESTDIR)/$(testsdir)/$$f; \
+		install -D -m 755 $(top_builddir)/$$f $(DESTDIR)/$(testsdir)/$$f; \
```
Comment (martinpitt) [inline Makefile.am]: Are these two `$(top_builddir)` actually necessary? I thought the cwd of make already was `$(builddir)`, and we only have one top-level Makefile. If they are necessary, then it needs to be added to the `[ -x .libs ]` above too.

File: Makefile.am [hunk 2]
```diff
@@ -6702,13 +6702,13 @@ tests += \
...
 	    fi; \
 	done
 	for f in $(TEST_DATA_FILES); do \
-	    install -D -m 644 $$f $(DESTDIR)/$(testsdir)/testdata/$${f#test/}; \
+	    install -D -m 644 $(top_srcdir)/$$f $(DESTDIR)/$(testsdir)/testdata/$${f#test/}; \
```
Comment (martinpitt) [inline Makefile.am:6725]: Argh, nice catch, thanks! Curious that `distcheck` didn't spot that.

File: src/shared/tests.c [hunk 3]
```diff
@@ -42,5 +42,9 @@ const char* get_exe_relative_testdata_dir(void) {
 
         assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
         assert_se(snprintf(testdir, sizeof(testdir), "%s/testdata", dirname(exedir)) > 0);
+        if (access(testdir, F_OK) < 0) {
+                fprintf(stderr,  "Test data directory '%s' does not exist, set $TEST_DIR\n", testdir);
+                exit(1);
```
Comment (evverx) [inline src/shared/tests.c]: `exit(1)` prevents all cleanup actions.

For example, `test_engine` creates the directory:
```c

        assert_se(runtime_dir = setup_fake_runtime_dir());

        /* prepare the test */
        assert_se(set_unit_path(TEST_DATA_DIR("")) >= 0);
```
I got many `/tmp/fake-xdg-runtime-*`-directories :-)

Maybe, we should call `TEST_DATA_DIR` before `setup_fake_runtime_dir`. As another option we can propagate error from `get_exe_relative_testdata_dir`

Comment (martinpitt) [reply src/shared/tests.c]: Or just write it to stderr to get it as first error message, and let the test fail by itself later?

Comment (keszybz) [reply src/shared/tests.c]: Yeah, doing `set_unit_path` before any "real" work seems best.

**Comment by martinpitt** [comment]
We could first check in dir_of_exe/testdata/, and then try dir_of_exe/../testdata/, so that it would work in the common case of builddir == srcdir. Then one only needs to set $TEST_DIR for builddir != srcdir. That should cover most of the "run test manually" cases and for the remainder we have an useful error message now.

File: src/shared/tests.c (hunk 3 above)
Comment (evverx) [reply src/shared/tests.c]: > and let the test fail by itself later?

our tests fail via `ABRT`. this prevents all cleanup operations too.

> Or just write it to stderr to get it as first error message

I think
```
$ ./libtool --mode=execute ./test-engine
Test data directory '/home/vagrant/systemd/.libs/testdata' does not exist, set $TEST_DIR
```
is better then
```
some stdout
...
Test data directory '/home/vagrant/systemd/.libs/testdata' does not exist, set $TEST_DIR
many lines
aborted
```

File: Makefile.am (hunk 1 above)
Comment (mbiebl) [reply Makefile.am]: I was not really necessy for the executables. It just seemed more symmetric to add it there as well.

Speaking of the executable, I wonder if we should use libtool here. E.g. taken from install-rootlibexecLTLIBRARIES:
```
$(LIBTOOL) $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=install $(INSTALL) $(INSTALL_STRIP_FLAG) $$list2 "$(DESTDIR)$(rootlibexecdir)";
```

The .libs directory is an implementation detail of libtool.

The full install rule 
```
install-rootlibexecLTLIBRARIES: $(rootlibexec_LTLIBRARIES)
	@$(NORMAL_INSTALL)
	@list='$(rootlibexec_LTLIBRARIES)'; test -n "$(rootlibexecdir)" || list=; \
	list2=; for p in $$list; do \
	  if test -f $$p; then \
	    list2="$$list2 $$p"; \
	  else :; fi; \
	done; \
	test -z "$$list2" || { \
	  echo " $(MKDIR_P) '$(DESTDIR)$(rootlibexecdir)'"; \
	  $(MKDIR_P) "$(DESTDIR)$(rootlibexecdir)" || exit 1; \
	  echo " $(LIBTOOL) $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=install $(INSTALL) $(INSTALL_STRIP_FLAG) $$list2 '$(DESTDIR)$(rootlibexecdir)'"; \
	  $(LIBTOOL) $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=install $(INSTALL) $(INSTALL_STRIP_FLAG) $$list2 "$(DESTDIR)$(rootlibexecdir)"; \
	}
```

File: src/shared/tests.c (hunk 3 above)
Comment (martinpitt) [reply src/shared/tests.c]: OK, agreed. I swapped around the setup of test dir and fake runtime dir in the four affected tests. Will push in a minute.

**Comment by martinpitt** [comment]
Re-pushed with additionally looking for the test data in the build tree layout, and with the above swapping of testdir/fake runtime dir setup to avoid leaving tmpfiles behind if the test dir cannot be found.

**Comment by keszybz** [comment]
I piled two commits on top (one replacing the last from @martinpitt). Please have a look. I used this branch to avoid multiplication of PRs, but please just drop my commits if you think the previous approach is better.

File: src/shared/tests.c [hunk 4]
```diff
@@ -41,6 +42,16 @@ const char* get_exe_relative_testdata_dir(void) {
...
+        if (path_startswith(exedir, ABS_BUILD_DIR))
+                return ABS_SRC_DIR "/test";
+
+        /* Try relative path, according to the install-test layout */
         assert_se(snprintf(testdir, sizeof(testdir), "%s/testdata", dirname(exedir)) > 0);
-        return testdir;
+        if (access(testdir, F_OK) >= 0)
+                return testdir;
+
+        fputs("ERROR: Cannot find testdata directory, set $SYSTEMD_TEST_DATA\n", stderr);
```
Comment (evverx) [inline src/shared/tests.c]: We show this message for default directories only. But
```
SYSTEMD_TEST_DATA=/some/testdata ./libtool --mode=execute ./test-engine
...
Test1: (Trivial)
error: org.freedesktop.systemd1.NoSuchUnit: Unit c.service not found.
Assertion 'r == 0' failed at src/test/test-engine.c:61, function main(). Aborting.
Aborted (core dumped)
```
Maybe, we should call `access(testdir, F_OK)` in `TEST_DATA_DIR`

I mean something like:
```
d = getenv(SYSTEMD_TEST_DATA)
if !d
     d = get_exe_relative_testdata_dir()
if access(d, F_OK) >= 0
     return d
...something went wrong here...
```

File: src/shared/tests.c [hunk 5]
```diff
@@ -41,10 +42,16 @@ const char* get_exe_relative_testdata_dir(void) {
         static char testdir[PATH_MAX];
 
         assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
+
+        /* Check if we're running from the builddir. If so, use the compiled in path. */
+        if (path_startswith(exedir, ABS_BUILD_DIR))
+                return ABS_SRC_DIR "/test";
```
Comment (martinpitt) [inline src/shared/tests.c]: Hah, clever! I like this.

File: src/shared/tests.c (hunk 4 above)
Comment (martinpitt) [reply src/shared/tests.c]: You mean setting S_T_D to a wrong directory? That seems like a weird corner case to me, but this can certainly grow an error message too. However, not the same one, as in that case S_T_D actually *is* set, so it would be confusing.

Comment (evverx) [reply src/shared/tests.c]: I mean setting S_D_T to a non-existent directory.

`S_D_T=totally-wrong-directory` will produce strange output. I think that's OK. (I have no idea how to check correctness of the S_D_T-contents :))

**Comment by martinpitt** [comment]
I added another commit on top which simplifies the structure/names and now also prints an error message for a non-existing $SYSTEMD_TEST_DATA. @evverx , do you mind reviewing this?

**Comment by keszybz** [comment]
Oops, we have a problem. `alloca` should not be called from within args of a function. It seems *all* uses of TEST_DATA_DIR are wrong.

**Comment by martinpitt** [comment]
@kesbyz: oh, I wasn't aware of that and it seems to work quite fine. Anyway, with the latest commit here the macro is not really doing much any more, I could just change get_testdata_dir() to take an optional subdir and append it in there (which works quite naturally). Do you want a new PR for that (as the strjoina() is already in master), or want me to push this on top of this one?

**Comment by keszybz** [comment]
Either way works for me. (I don't know in what circumstances `alloca` inside of a function call would cause problems. It might depend on the architecture, and the compiler, and the optimization level. I don't think we've ever seen actual problems, but the standard is pretty clear that this is not allowed.)

**Comment by martinpitt** [comment]
I lobbed another commit on top which drops the macro and thus avoids the `alloca()`. This required some massaging of `get_testdata_dir()`, I restructured this a bit to be more readable. This now covers all three cases (installed test, build tree test, and arbitrary location with the env var).

File: src/shared/tests.c [hunk 6]
```diff
@@ -36,33 +36,37 @@ char* setup_fake_runtime_dir(void) {
...
-                if (access(env, F_OK) >= 0)
-                        return env;
-
-                fputs("ERROR: $SYSTEMD_TEST_DATA directory does not exist\n", stderr);
-                exit(1);
+                if (access(env, F_OK) < 0) {
+                        fputs("ERROR: $SYSTEMD_TEST_DATA directory does not exist\n", stderr);
+                        exit(1);
+                }
+                strncpy(testdir, env, sizeof(testdir) - 1);
```
Comment (keszybz) [inline src/shared/tests.c:52]: This leaves `testdir` unterminated, potentially.

Comment (martinpitt) [reply src/shared/tests.c:52]: Argh, fixed.

Comment (keszybz) [reply src/shared/tests.c:52]: Hm, wasn't this supposed to be changed?

Comment (martinpitt) [reply src/shared/tests.c:52]: I did after the strncpy() below; but indeed this wouldn't suffice as not terminating it here would confuse the second strncpy()'s strlen() call. Sorry about that. (Then again, there are probably easier ways to do harm to yourself than trying to inject bogus SYSTEMD_TEST_DATA values, but let's be correct here indeed.)

Also, the termination shouldn't be necessary in the first place as static variables (like global ones) should get nulled at program start. But efficiency doesn't matter here. I pushed again to move the null termination further up.

**Comment by martinpitt** [comment]
Hang on, needs a rebase for the new test-journal-importer. Running local tests..

**Comment by martinpitt** [comment]
Rebased to current master, should be good now.

File: src/shared/tests.c (hunk 6 above)
Comment (keszybz) [reply src/shared/tests.c:52]: A, right, a static variable. You just have to love C.

**Comment by keszybz** [comment]
Looks good. I pushed a change to simplify the environment definition, we don't need to add those variables now.

File: Makefile.am [hunk 7]
```diff
@@ -6615,7 +6614,7 @@ valgrind-tests: $(TESTS)
 		if $(LIBTOOL) --mode=execute file $$f | grep -q shell; then \
 		echo -e "$${x}Skipping non-binary $$f"; else \
 		echo -e "$${x}Running $$f"; \
-		$(AM_TESTS_ENVIRONMENT) $(LIBTOOL) --mode=execute valgrind -q --leak-check=full --max-stackframe=5242880 --error-exitcode=55 $(builddir)/$$f ; fi; \
+		$(LIBTOOL) --mode=execute valgrind -q --leak-check=full --max-stackframe=5242880 --error-exitcode=55 $(builddir)/$$f ; fi; \
```
Comment (evverx) [inline Makefile.am]: not sure we should remove `AM_TEST_ENVIRONMENT`. It makes sense to set the environment variables for all the tests (under `valgrind` or not).

Comment (martinpitt) [reply Makefile.am]: Yes, this was an explicit fix, we also need it for the other variables -- otherwise we test against the *installed* systemd binaries, but we want those from the build tree. @keszybz, I force-pushed your commit to only keep the first hunk, and revert the invocation of valgrind (i. e. the second hunk).

**Comment by evverx** [comment]
For the record
```sh
sudo make install install-tests
cd /usr/lib/systemd/tests
./test-libsystemd-sym
./test-libsystemd-sym: /lib/i386-linux-gnu/libsystemd.so.0: version `LIBSYSTEMD_233' not found (required by ./test-libsystemd-sym)
```
Some tests ignore `SYSTEMD_TEST_DATA`
```sh
$ SYSTEMD_TEST_DATA=/wow ./libtool --mode=execute ltrace ./test-nss |& grep -i dlopen
dlopen("./.libs/libnss_myhostname.so.2", 4097)   = 0x80869118
dlopen("./.libs/libnss_resolve.so.2", 4097)      = 0x8086b1e8
```
(but we don't install `./test-nss` into `/usr/lib/systemd/tests`. Maybe, this doesn't really matter)

I think we can fix these issues later

**Comment by martinpitt** [comment]
@evverx: I don' think there's much we can do about the first issue. That's exactly the same as copying e. g. systemd-networkd from the build tree into the installed system, since we moved to the shared library you need to install everything or nothing (or at least specify the path to the library, which is what I do for testing). Good  point about test-nss though: I wanted to adjust this to install the manual and unsafe tests too (into subdirs), will make test-nss use `get_testdata_dir()` then. Thanks for merging!

**Comment by evverx** [comment]
> I don' think there's much we can do about the first issue

@martinpitt , maybe, `make install-tests` shouldn't install `test-libsystemd-sym`, `test-libudev-sym`. We can treat these tests as "internal" or so

**Comment by martinpitt** [comment]
@evverx : Pretty much all of the tests need the shared library; I don't see how the -sym ones are special?

**Comment by evverx** [comment]
@martinpitt , well, this is my bad. Actually,
```
$ ldd test-libsystemd-sym
...
libsystemd.so.0 => /lib/i386-linux-gnu/libsystemd.so.0
```
But I installed `libsystemd.so` to `/lib/` somehow. i.e.
```
LD_PRELOAD=/lib/libsystemd.so.0 ./test-libsystemd-sym
```
works fine

I should use `--with-rootlibdir=/lib/i386-linux-gnu` (instead of `--with-rootlibdir=/lib`)

I've already [wrote](https://github.com/systemd/systemd/pull/5339#issuecomment-279735148):
> Not sure what to do about it. Maybe, we should simply add the section "how to properly make install-tests" to the README

:-)
